CodeCoverageQualityGate/CodeCoverageQualityService.ts

```diff
@@ -287,9 +322,181 @@ export class CodeCoverageQualityService {
+                    break;
+                default:
+                    // User will never come here
+                    tl.error(`The build status should be unstable or failed instead of ${this.buildStatus}`);
+            }
+        }
+        tl.debug(`Quality Gate ${qualityGateResult.status} - ${this.type} code coverage: ${qualityGateResult.codeCoverage} (${coveredLines}/${coverableLines}) - Threshold: ${this.threshold}%`);
+        return qualityGateResult;
+    }
+
+    private getOverallCodeCoverage = (coberturaContentString: string): void => {
+        const saxParser = sax.parser(true);
+        saxParser.onopentag = (node) => {
+            if (node.name == 'coverage') {
+                this.overallCoveredLines = parseInt(<string>node.attributes['lines-covered']);
+                this.overallCoverabledLines = parseInt(<string>node.attributes['lines-valid']);
+            }
+        }
+
+        saxParser.onend = () => {
+            // do nothing
+        };
+
+        saxParser.onerror = (e) => {
+            tl.warning('Failed to parse cobertura code coverage report content.');
+            console.error(e);
+        };
+
+        saxParser.write(coberturaContentString).close();
+    }
+
+
     private getQualityGateIdentification() {
         const referencePipeline: string = this.originalReferencePipelineName ? ", Reference pipeline: " + this.originalReferencePipelineName : "";
         const referenceBuild: string = this.originalReferenceBuildNumber ? ", Reference build: " + this.originalReferenceBuildNumber : "";
         return "Type: " + this.type + ", Threshold: " + this.threshold + referencePipeline + referenceBuild;
     }
+
+    private getModifiedCodeCoverage = (referenceBuildCoverageReport: string, currentBuildCoverageReport: string): void => {
+        const currentBuildCodeFileContainer: CodeFile[] = [];
+        this.getCoverageDataFromReport(currentBuildCoverageReport, currentBuildCodeFileContainer);
+        const referenceBuildCodeFileContainer: CodeFile[] = [];
+        this.getCoverageDataFromReport(referenceBuildCoverageReport, referenceBuildCodeFileContainer);
+
+        this.calculateModifiedCodeLines(currentBuildCodeFileContainer, referenceBuildCodeFileContainer);
+    }
+
+    private getCoverageDataFromReport = (reportContent: string, container: CodeFile[]): void => {
+        const saxParser = sax.parser(true);
+        let currentCodeFile : CodeFile = {
+            fileIdentity: '',
+            codeLines: []
+        }
+
+        saxParser.onopentag = (node) => {
+            if (node.name == 'class') {
+                currentCodeFile = {
+                    fileIdentity: <string>node.attributes.filename + ':' + <string>node.attributes.name,
+                    codeLines: []
+                }
+            }
+            if (node.name == 'line') {
+                let line: CodeLine = {
+                    lineNumber: parseInt(<string>node.attributes.number),
```

Comment:
Is it possible to encounter invalid attribute values here (e.g. attribute doesn't exist), or are they handled during the transformation to the Cobertura report, such as by throwing an error or exception? I asked this question because in all following implementation we assume each "CodeLine" will have the three attributes for sure.

Reply:
According to cobertura.xsl,  the cobertura report have the three attributes for sure. Even if some values are missing, the attribute still exist. For example: if linehash is missing, the attribute will look like **hash=""**. As your concer, we have two ways to handle this and i prefer the second:
1. We could set a default value for each attribute when atrribute value is empty string.
2. Set the quality gate failed, because when this situtation happen, the modified code coverage is not accurate.

But in my opinion, in normal situtaion(parsing cobertura report which generated by parasoft report), it is impossible meet a attribute which have a invalid value.

Reply:
Even if it's impossible to get invalid values in the report through a normal workflow, we should add code to guard against accidental file modifications. Please proceed with the second option to validate the values of attributes before setting them. 
Additionally, we might consider setting the task of the quality gate to "skipped," if such an option exists, instead of failing the quality gate. Please share your thoughts on this, @rwang-parasoft. 
In my opinion, that doesn't count as a failed situation but rather as a situation where I can't evaluate the quality gate, so I should skip doing it.